### ISSUE TITLE
[Question] Why is there no initial model sync in data parallel training on GPU.

### ISSUE THREAD

**Issue by cicirori** [pytorch/xla#3216]
## ❓ Questions and Help

Hi, I am using test_train_mp_minist.py to test distributed training on GPU. But during this process I noticed that unlike DDP, torch xla doesn't seem to synchronize the model parameters at the beginning. I did not see the broadcast call in the NCCL log, and to verify if the model parameter synchronization was done somewhere, I performed an experiment: using two cards for distributed training, with the model on one card using random initialization and the other card doing nothing, and printing the model parameters after the forward part of the first iteration, I found that they were still inconsistent.
I wonder why, since inconsistent initial parameters for different workers should be a relatively common behavior in distributed training. For example, all cards are loaded with the same model structure, but only rank 0 holds model checkpoints that maybe from previous training.
I'm wondering if I'm doing something wrong. If not, do you have any plans to support model synchronization in the initial stages of distributed training.

**Comment by miaoshasha** [comment]
I thought that script is for running on TPUs? can it run on GPUs as well? @miladm

**Comment by jysohn23** [comment]
Hi @cicirori thanks for raising the discrepancy. As you describe, when you wrap your model with DDP it begins by synchronizing the buffers and params of the model: https://github.com/pytorch/pytorch/blob/ab1d879b33bae968e22e098b0e6aa01af821efe2/torch/nn/parallel/distributed.py#L603

For DataParallel mode (multithreading not multiprocessing) we used to do similar where we instantiate / copy model with same weights. https://github.com/pytorch/xla/blob/73a3937a1da542c4fd0b93097abe797be94c567a/torch_xla/distributed/data_parallel.py#L58 

However, with multiprocessing standard practice is for the user to have the same seed for all workers since we're in replicated training and want each worker to do the same thing. If for some reason you can't initialize same weights, you could manually do the syncing yourself with:
https://github.com/pytorch/xla/blob/73a3937a1da542c4fd0b93097abe797be94c567a/torch_xla/core/xla_model.py#L529

@JackCaoG can consider if it's best for PyTorch / XLA to carry a wrapper/util that does that for you.

**Comment by jeffhataws** [comment]
@miaoshasha , @JackCaoG , when we will have this automatic sync in PyTorch-XLA? Would you also update documentation to indicate this "quirk" until there's an automatic sync implementation? We also see this same issue with HuggingFace BERT multi-worker training, where there's difference between DDP and XMP spawn on GPU due to this missing sync in XLA. We also found that keeping the same seed in all workers help close the gap between PyTorch and PyTorch-XLA. Additionally, we found that calling model.init_weights() function available in HuggingFace BERT also cause initial weights to become identical. This mechanism is not obvious to me, but it is the same on PyTorch as PyTorch-XLA.
